modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -1801,4 +1842,178 @@ private boolean enterBusy() {
     private void leaveBusy() {
         opsLock.readLock().unlock();
     }
+
+    /**
+     * Registers metrics to measure durations of service methods.
+     *
+     * @param srvc Service for invocations measurement.
+     * @param srvcName Name of {@code srvc}.
+     */
+    private void registerMetrics(Service srvc, String srvcName) {
+        getInterfaces(srvc.getClass()).stream().map(Class::getMethods).flatMap(Arrays::stream)
+            .filter(mtd -> !isMetricIgnoredFor(mtd.getDeclaringClass()))
+            .forEach(mtd -> {
+                // All metrics for current service.
+                Map<String, MethodHistogramHolder> srvcHistograms =
+                    invocationHistograms.computeIfAbsent(srvcName, name -> new HashMap<>(1));
+
+                // Histograms for this method name.
+                MethodHistogramHolder mtdHistograms =
+                    srvcHistograms.computeIfAbsent(mtd.getName(), mdtName -> new MethodHistogramHolder());
+
+                mtdHistograms.addIfAbsent(mtd, () -> createHistogram(srvcName, mtd));
+            });
+    }
+
+    /**
+     * Removes metrics for service {@code srvcName}.
+     *
+     * @param srvcName Service name.
+     */
+    private void unregisterMetrics(String srvcName) {
+        ctx.metric().remove(serviceMetricRegistryName(srvcName));
+
+        invocationHistograms.remove(srvcName);
+    }
+
+    /**
+     * Creates histogram for service method. If exist,s considers one or several argument types has same name but
+     * different package and tries to extend metric name with abbreviation of java package name.
+     *
+     * @param srvcName Service name.
+     * @param method Method to measure.
+     * @return Histogram of service method timings.
+     */
+    HistogramMetricImpl createHistogram(String srvcName, Method method) {
+        MetricRegistry metricRegistry = ctx.metric().registry(serviceMetricRegistryName(srvcName));
+
+        HistogramMetricImpl histogram = null;
+
+        // Find/create histogram.
+        for (int i = 0; i <= MAX_ABBREVIATE_NAME_LVL; ++i) {
+            String methodMetricName = methodMetricName(method, i);
+
+            synchronized (metricRegistry) {
```

Comment:
@ivandasch , there is a gap between
`metricRegistry.findMetric(methodMetricName)`
and
`metricRegistry.histogram(methodMetricName, DEFAULT_INVOCATION_BOUNDS,
                        DESCRIPTION_OF_INVOCATION_METRIC)`

It is important. If something creates metric after `metricRegistry.findMetric(methodMetricName)`, it would cause incorrent metric naming or NPE.

Reply:
In code of method `MetricRegistry#histogram` I see clearly, that if a metric with same name exists, this method returns the old metric with bounds reset. If this is not OK, I suggest to rewrite a little bit `MetricRegistry#histogram`  or create specialized method. This task can be done without blocking code, I presume.

Reply:
In this case we need to skip metric if it exists and described in the javadoc: 

```
Creates histogram for service method. If exists, considers one or several argument types has same name but
     * different package and tries to extend metric name with abbreviation of java package name.
```

Please see `NamingService` and `GridServiceMetricsTest.testMetricNaming()`

Reply:
Ok, It seems to be a not a problem, because of one registry per service approach. Seems that we can leave blocking code as is